### ISSUE TITLE
iOS: Http request hangs the callback function

### ISSUE THREAD

**Issue by xiehaom7** [transistorsoft/cordova-plugin-background-fetch#107]
## Your Environment
* Plugin version: 5.3.0
* Platform: iOS
* OS version: 12.2 beta 3
* Device manufacturer / model: iPhone 6
* Cordova version (`cordova -v`): 7.1.0
* Cordova platform version (`cordova platform ls`): 4.5.5
* Plugin config
com-sarriaroman-photoviewer 1.1.18 "PhotoViewer"
cordova-fabric-plugin 1.1.14-dev "cordova-fabric-plugin"
cordova-plugin-actionsheet 2.3.3 "ActionSheet"
cordova-plugin-add-swift-support 1.7.2 "AddSwiftSupport"
cordova-plugin-android-permissions 1.0.0 "Permissions"
cordova-plugin-app-event 1.2.1 "Application Events"
cordova-plugin-app-version 0.1.9 "AppVersion"
cordova-plugin-background-fetch 5.3.0 "CDVBackgroundFetch"
cordova-plugin-badge 0.8.8 "Badge"
cordova-plugin-battery-status 2.0.2 "Battery"
cordova-plugin-camera 4.0.3 "Camera"
cordova-plugin-device 1.1.7 "Device"
cordova-plugin-dialogs 2.0.1 "Notification"
cordova-plugin-email-composer 0.8.15 "EmailComposer"
cordova-plugin-file 6.0.1 "File"
cordova-plugin-file-opener2 2.1.3 "File Opener2"
cordova-plugin-file-transfer 1.7.1 "File Transfer"
cordova-plugin-googlemaps 2.5.0 "cordova-plugin-googlemaps"
cordova-plugin-inappbrowser 3.0.0 "InAppBrowser"
cordova-plugin-ionic 5.2.10 "cordova-plugin-ionic"
cordova-plugin-ionic-keyboard 2.1.3 "cordova-plugin-ionic-keyboard"
cordova-plugin-ionic-webview 4.0.0 "cordova-plugin-ionic-webview"
cordova-plugin-local-notification 0.9.0-beta.2 "LocalNotification"
cordova-plugin-mauron85-background-geolocation 2.3.6 "CDVBackgroundGeolocation"
cordova-plugin-nativeaudio 3.0.9 "Cordova Native Audio"
cordova-plugin-network-information 1.3.4 "Network Information"
cordova-plugin-printer 0.7.3 "Printer"
cordova-plugin-splashscreen 5.0.2 "Splashscreen"
cordova-plugin-statusbar 2.4.2 "StatusBar"
cordova-plugin-whitelist 1.3.3 "Whitelist"
cordova-plugin-x-socialsharing 5.4.4 "SocialSharing"
cordova-sqlite-storage 2.6.0 "Cordova sqlite storage plugin"
es6-promise-plugin 4.2.2 "Promise"
mx.ferreyra.callnumber 0.0.2 "Cordova Call Number Plugin"
phonegap-plugin-mobile-accessibility 1.0.5-dev "Mobile Accessibility"
uk.co.workingedge.phonegap.plugin.launchnavigator 4.2.2 "Launch Navigator"

## Expected Behavior

The request should be resolved.

## Actual Behavior

Seems the request hangs the process without any error. Any code in the Promise.then is not executed. Without http request, there is no problem to do local storage or local notification.

## Steps to Reproduce

```javascript
const fetchCallbackFn = () => {
    console.log("fetch triggered ***********");
    this.awsCredential.sendRequest(`url`)
        .then(res => {
             console.log(res);
             console.log("fetch finished ***********");
             this.backgroundFetch.finish();
         });
    };
BackgroundFetch.configure(fetchCallbackFn,
    e => console.log("fetch error: ", e),
     { minimumFetchInterval: 15 });
```

## Context

We are trying to fetching updated information when app is in background mode 

## Debug logs

```
2019-02-26 17:44:50.598045-0600 [2036:419852] CDVBackgroundFetch AppDelegate received fetch event
2019-02-26 17:44:50.598162-0600 [2036:419852] [TSBackgroundFetch performFetchWithCompletionHandler]
2019-02-26 17:44:50.598204-0600 [2036:419852] - CDVBackgroundFetch Rx Fetch Event
2019-02-26 17:44:50.672614-0600 [2036:419852] fetch triggered ***********

// simulate fetch again

2019-02-26 17:46:41.214401-0600 [2036:419852] CDVBackgroundFetch AppDelegate received fetch event
2019-02-26 17:46:41.214516-0600 2036:419852] [TSBackgroundFetch performFetchWithCompletionHandler]
2019-02-26 17:46:41.214880-0600 [2036:419852] Warning: Application delegate received call to -application:performFetchWithCompletionHandler: but the completion handler was never called.
2019-02-26 17:46:41.214982-0600 [2036:419852] - CDVBackgroundFetch Rx Fetch Event
2019-02-26 17:46:41.358634-0600 [2036:419852] fetch triggered ***********
```

**Comment by christocracy** [comment]
Try `await`ing the request.
```javascript
const fetchCallbackFn = async () => {
    console.log("fetch triggered ***********");
    let res = await this.awsCredential.sendRequest(`url`);
    console.log(res);
    console.log("fetch finished ***********");
    this.backgroundFetch.finish(); 
};```

**Comment by xiehaom7** [comment]
Thanks! I will test this tomorrow.

**Comment by xiehaom7** [comment]
I tested the async/await, but that doesn't help. 

I noticed that under another issue, you used http plugin to test. So we tried the http plugin and that works. Unfortunately, I'm using Amazon's aws Javascript SDK to reach their services and I assume they just use javascript's built-in APIs. Seems javascript request are blocked by iOS when app/webview is in background. Do you know any solution for this? I'm really appreciated for your help.

**Comment by christocracy** [comment]
No idea, sorry.

**Comment by christocracy** [comment]
Are you watching the XCode logs?  Did you white-list the Amazon url in your app's `.plist` file? 

![](https://dl.dropbox.com/s/4xe7olz35k2obve/Screenshot%202019-02-27%2013.19.26.png?dl=1)

**Comment by xiehaom7** [comment]
I checked the setting. It looks exactly like your screenshot. In addition, I don't have problem connecting to Amazon with their API in foreground.

Do I still need to add the exception domain? Sorry for any stupid questions. I have limited knowledge about iOS development. Thanks.

**Comment by christocracy** [comment]
`Allow Arbitrary Loads` allows outgoing HTTP requests to anywhere.

**Comment by xiehaom7** [comment]
Yes. So I guess this is not the culprit.
